core/src/main/java/org/p2p/core/network/ConnectionManager.kt

```diff
@@ -0,0 +1,83 @@
+package org.p2p.core.network
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
+class ConnectionManager(context: Context) {
+    @Deprecated("Use flow instead", ReplaceWith("connectionStatus"))
+    interface Listener {
+        fun onConnectionChange()
+    }
+
+    private val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+    private val _connectionStatus = MutableStateFlow(true)
+    val connectionStatus = _connectionStatus.asStateFlow()
+
+    @Deprecated("Use flow instead", ReplaceWith("connectionStatus"))
+    var listener: Listener? = null
+
+    private var callback = ConnectionStatusCallback()
+
+    init {
+        connectivityManager.registerNetworkCallback(
+            NetworkRequest.Builder()
+                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                .addCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
+                .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
+                .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
+                .build(),
+            callback
+        )
+
+        notifyHasConnection(
+            connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork).hasValidatedInternet()
+        )
+    }
+
+    private fun notifyHasConnection(hasConnection: Boolean) {
+        listener?.onConnectionChange()
+        _connectionStatus.tryEmit(hasConnection)
+    }
+
+    private fun NetworkCapabilities?.hasValidatedInternet(): Boolean {
+        if (this == null) return false
+
+        return hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+            && hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
```

Comment:
I think we leave && on the first line instead of putting it on the new line

Reply:
1. It won't fit to 120 symbols in row (static importing leads to implicitness)
2. new line has better readability